plugins/ssr/ssrfwd.ml

```diff
@@ -89,26 +89,32 @@ let combineCG t1 t2 f g = match t1, t2 with
  | _, (_, (_, None)) -> anomaly "have: mixed C-G constr"
  | _ -> anomaly "have: mixed G-C constr"
 
-let basecuttac name t =
-  let open Proofview.Notations in
-  Ssrcommon.tacMK_SSR_CONST name >>= fun hd ->
-  let t = EConstr.mkApp (hd, [|t|]) in
-  Ssrcommon.tacTYPEOF t >>= fun _ty ->
-  Tactics.apply t
 
-let basesufftac t =
+ type cut_kind = Have | Suff
+ let basecuttac k c =
   let open Proofview.Notations in
-  Ssrcommon.tacMK_SSR_CONST "ssr_suff" >>= fun hd ->
-  let t = EConstr.mkApp (hd, [|t|]) in
-  Ssrcommon.tacTYPEOF t >>= fun _ty ->
-  Ssrcommon.applyn ~with_evars:true 3 t
+  Proofview.Goal.enter begin fun gl ->
+    let env = Proofview.Goal.env gl in
+    let sigma = Tacmach.New.project gl in
+    let concl = Proofview.Goal.concl gl in
+    match Typing.sort_of env sigma c with
+    | exception e when CErrors.noncritical e ->
+      let _, info = Exninfo.capture e in
+      Tacticals.New.tclZEROMSG ~info (str "Not a proposition or a type.")
+    | sigma, s ->
+      let r = Sorts.relevance_of_sort s in
+      let (sigma, f) = Evarutil.new_evar env sigma (EConstr.mkArrow c r concl) in
+      let g = Proofview_monad.with_empty_state (fst @@ EConstr.destEvar sigma f) in
+      Proofview.tclTHEN (Proofview.Unsafe.tclEVARS sigma) (Tactics.apply f)
+      <*>
+      Proofview.Unsafe.tclGETGOALS >>= begin fun gl ->
+        match k with
+        | Suff -> Proofview.Unsafe.tclSETGOALS (g :: gl)
+        | Have -> Proofview.Unsafe.tclSETGOALS (gl @ [g])
```

Comment:
@ppedrot suggestion on how to use this API are welcome. How on earth do I need to use `Unsafe` to declare a goal?

Reply:
Because unsafe is where the goal primitives are

Reply:
So either it is misnamed, or the primitives are in the wrong place.

Reply:
I agree that the `Unsafe` naming is not very well chosen, it would be better to call it something like `LowLevel`. That said, in theory you should be able to implement this without resorting to the low-level API and using refine instead. Declaring an evar in a refine block implicitly transforms it into a goal, and then you can cycle / swap the hell out of the result.